GitUI/Editor/RebaseTodoHighlightingStrategy.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System.Collections.Generic;
+using System.Drawing;
+using GitCommands;
+using ICSharpCode.TextEditor;
+using ICSharpCode.TextEditor.Document;
+using JetBrains.Annotations;
+
+namespace GitUI.Editor
+{
+    internal sealed class RebaseTodoHighlightingStrategy : GitHighlightingStrategyBase
+    {
+        /*
+        Commands:
+        p, pick = use commit
+        r, reword = use commit, but edit the commit message
+        e, edit = use commit, but stop for amending
+        s, squash = use commit, but meld into previous commit
+        f, fixup = like "squash", but discard this commit's log message
+        x, exec = run command (the rest of the line) using shell
+        d, drop = remove commit
+        */
+
+        private static readonly Dictionary<char, (string longForm, HighlightColor color)> _commandByFirstChar = new Dictionary<char, (string longForm, HighlightColor color)>
+        {
+            { 'p', ("pick", new HighlightColor(Color.Black, bold: true, italic: false)) },
+            { 'r', ("reword", new HighlightColor(Color.Purple, bold: true, italic: false)) },
+            { 'e', ("edit", new HighlightColor(Color.Black, bold: true, italic: false)) },
+            { 's', ("squash", new HighlightColor(Color.DarkBlue, bold: true, italic: false)) },
+            { 'f', ("fixup", new HighlightColor(Color.LightCoral, bold: true, italic: false)) },
+            { 'x', ("exec", new HighlightColor(Color.Black, bold: true, italic: false)) },
+            { 'd', ("drop", new HighlightColor(Color.Red, bold: true, italic: false)) }
+        };
+
+        public RebaseTodoHighlightingStrategy([NotNull] GitModule module)
+            : base("GitRebaseTodo", module)
+        {
+        }
+
+        protected override void MarkTokens(IDocument document, IList<LineSegment> lines)
+        {
+            foreach (var line in lines)
+            {
+                if (!TryHighlightComment(document, line) &&
+                    !TryHighlightInteractiveRebaseCommand(document, line))
+                {
+                    line.Words = new List<TextWord>(capacity: 1)
+                        { new TextWord(document, line, 0, line.Length, ColorNormal, hasDefaultColor: true) };
+                }
+
+                document.RequestUpdate(
+                    new TextAreaUpdate(TextAreaUpdateType.SingleLine, line.LineNumber));
+            }
+        }
+
+        private enum State
+        {
+            Command,
+            SpacesAfterCommand,
+            Id
+        }
+
+        private static bool TryHighlightInteractiveRebaseCommand(IDocument document, LineSegment line)
+        {
+            if (line.Length < 1)
+            {
+                return false;
+            }
+
+            var c = document.GetCharAt(line.Offset);
```

Comment:
This method would take a lot less code using Regex, but the text editor doesn't make content available as strings by default. You can force one out of it, but it's made hard for a reason. You want these functions to be as light as possible, and allocating strings can introduce GC pauses.

This approach uses hardly any allocations, other than the list of words on the line.